File: clrs/7/quicksort.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define SWAP(a, b, t)   \
    do {                \
        t temp = a;     \
        a = b;          \
        b = temp;       \
    } while (0)         \

int
partition(int buf[], int low, int high)
{
    int i, j, pivot;

    pivot = buf[high];
    i = low;
    for (j = low; j < high; j++)
        if (buf[j] <= pivot) {
            SWAP(buf[i], buf[j], int);
            i++;
        }
    SWAP(buf[i], buf[high], int);

    return i;
}

int
hoare_partition1(int buf[], int low, int high)
{
    int pivot = buf[low];

    while (1) {
        while (buf[high] > pivot)
            high--;
        while (buf[low] < pivot)
            low++;
        if (low < high) {
            SWAP(buf[low], buf[high], int);
            low++;
            high--;
        } else
            return high;
    }
}

int
hoare_partition2(int buf[], int low, int high)
{
    int pivot, i, j;

    pivot = buf[low];
    i = low - 1;
    j = high + 1;

    while (1) {
        do {
            j--;
        } while (buf[j] > pivot);

        do {
            i++;
        } while (buf[i] < pivot);

        if (i < j)
            SWAP(buf[i], buf[j], int);
        else {
            return j;
        }
    }
}


int
hoare_partition3(int buf[], int low, int high)
{
    int pivot = buf[low];
    int i = low, j = high;
    while (i < j) {
        while (i < j && buf[j] >= pivot)
            j--;
        buf[i] = buf[j];
        while (i < j && buf[i] <= pivot)
            i++;
        buf[j] = buf[i];
    }
    buf[i] = pivot;
    return i;

}

void
quicksort(int buf[], int low, int high)
{
    int mid, temp;
    int i;

    if (low < high) {
        mid = hoare_partition3(buf, low, high);
        quicksort(buf, low, mid - 1);
        quicksort(buf, mid+1, high);
    }
}

File: clrs/7/quicksort.py
#!/usr/bin/env python
from randata import getrandata
from timing import timing
import random

def quicksort(arr, low, high):
    if low < high:
        mid = partition(arr, low, high)
        quicksort(arr, low, mid - 1)
        quicksort(arr, mid + 1, high)

def partition(arr, low, high):
    pivot = arr[low]
    i = low
    for j in range(low + 1, high + 1):
        if arr[j] < pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
    arr[i], arr[low] = arr[low], arr[i]
    return i

def randomized_partition(arr, low, high):
    pivot = random.randint(low, high)
    arr[pivot], arr[low] = arr[low], arr[pivot]
    return partition(arr, low, high)

def randomized_quicksort(arr, low, high):
    if low < high:
        mid = randomized_partition(arr, low, high)
        randomized_quicksort(arr, low, mid - 1)
        randomized_quicksort(arr, mid + 1, high)


if __name__ == '__main__':
    arr = getrandata(20)
    print(arr)
    
    randomized_quicksort(arr, 0, len(arr) - 1)
    print(arr)


        
